File: relay/python/aether_relay/noise.py
"""Minimal Noise-like upgrade using X25519 + HKDF + ChaCha20-Poly1305.

NOTE: This is a lightweight, interoperable encryption layer and not a
full Noise XX implementation. It is intended to provide confidentiality
with a clear upgrade path to a full Noise framework.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Literal

from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.ciphers.aead import ChaCha20Poly1305
from cryptography.hazmat.primitives.kdf.hkdf import HKDF
from cryptography.hazmat.primitives.asymmetric.x25519 import (
    X25519PrivateKey,
    X25519PublicKey,
)


def generate_keypair() -> tuple[bytes, bytes]:
    private_key = X25519PrivateKey.generate()
    public_key = private_key.public_key()
    return (
        private_key.private_bytes_raw(),
        public_key.public_bytes_raw(),
    )


def derive_shared_key(private_key: bytes, peer_public_key: bytes) -> bytes:
    priv = X25519PrivateKey.from_private_bytes(private_key)
    pub = X25519PublicKey.from_public_bytes(peer_public_key)
    shared = priv.exchange(pub)
    hkdf = HKDF(
        algorithm=hashes.SHA256(),
        length=32,
        salt=None,
        info=b"aether-noise",
    )
    return hkdf.derive(shared)


@dataclass
class NoiseSession:
    key: bytes
    send_counter: int = 0

    def encrypt(self, plaintext: bytes) -> bytes:
        nonce = self._nonce(self.send_counter)
        self.send_counter += 1
        cipher = ChaCha20Poly1305(self.key)
        ciphertext = cipher.encrypt(nonce, plaintext, None)
        return self._counter_prefix(self.send_counter - 1) + ciphertext

    def decrypt(self, payload: bytes) -> bytes:
        if len(payload) < 8:
            raise ValueError("noise payload too short")
        counter = int.from_bytes(payload[:8], "big")
        nonce = self._nonce(counter)
        cipher = ChaCha20Poly1305(self.key)
        return cipher.decrypt(nonce, payload[8:], None)

    @staticmethod
    def _nonce(counter: int) -> bytes:
        return b"\x00\x00\x00\x00" + counter.to_bytes(8, "big")

    @staticmethod
    def _counter_prefix(counter: int) -> bytes:
        return counter.to_bytes(8, "big")

File: sdk/typescript/src/noise.ts
import { chacha20poly1305 } from "@noble/ciphers/chacha";
import { hkdf } from "@noble/hashes/hkdf";
import { sha256 } from "@noble/hashes/sha256";
import { x25519 } from "@noble/curves/ed25519";

export function generateKeypair(): { privateKey: Uint8Array; publicKey: Uint8Array } {
  const privateKey = x25519.utils.randomPrivateKey();
  const publicKey = x25519.getPublicKey(privateKey);
  return { privateKey, publicKey };
}

export function deriveSharedKey(privateKey: Uint8Array, peerPublicKey: Uint8Array): Uint8Array {
  const shared = x25519.getSharedSecret(privateKey, peerPublicKey);
  return hkdf(sha256, shared, undefined, "aether-noise", 32);
}

export class NoiseSession {
  private key: Uint8Array;
  private sendCounter = 0;

  constructor(key: Uint8Array) {
    this.key = key;
  }

  encrypt(plaintext: Uint8Array): Uint8Array {
    const nonce = this.nonce(this.sendCounter);
    const cipher = chacha20poly1305(this.key, nonce);
    const ciphertext = cipher.encrypt(plaintext);
    const counter = this.counterPrefix(this.sendCounter);
    this.sendCounter += 1;
    const out = new Uint8Array(counter.length + ciphertext.length);
    out.set(counter, 0);
    out.set(ciphertext, counter.length);
    return out;
  }

  decrypt(payload: Uint8Array): Uint8Array {
    if (payload.length < 8) {
      throw new Error("noise payload too short");
    }
    const counter = bytesToNumber(payload.slice(0, 8));
    const nonce = this.nonce(counter);
    const cipher = chacha20poly1305(this.key, nonce);
    return cipher.decrypt(payload.slice(8));
  }

  private nonce(counter: number): Uint8Array {
    const nonce = new Uint8Array(12);
    nonce.set(numberToBytes(counter), 4);
    return nonce;
  }

  private counterPrefix(counter: number): Uint8Array {
    return numberToBytes(counter);
  }
}

function numberToBytes(value: number): Uint8Array {
  const out = new Uint8Array(8);
  let v = BigInt(value);
  for (let i = 7; i >= 0; i -= 1) {
    out[i] = Number(v & 0xffn);
    v >>= 8n;
  }
  return out;
}

function bytesToNumber(bytes: Uint8Array): number {
  let v = 0n;
  for (const b of bytes) {
    v = (v << 8n) + BigInt(b);
  }
  return Number(v);
}
